clean_air/data/storage.py

```diff
@@ -198,3 +181,211 @@ def put(self, item: AURNSite) -> None:
+        # TODO, lazy loading, some kind of closure or DataSet subclass (LazyLoadingS3DataSet?)?
+        s3_key = self._generate_s3_key(dataset_id)
+        cache_dir_path = self._generate_cache_dir_path(dataset_id)
+
+        try:
+            metadata = self._metadata_store.get(dataset_id)
+            self._fs.get(s3_key, str(cache_dir_path), recursive=True)
+        except PermissionError:
+            msg = f"PermissionError when accessing {s3_key}."
+            msg += " Object may not exist, or you may have incorrect/misconfigured credentials"
+            raise DataStoreException(msg)
+        datafile_paths = self._get_datafile_paths(cache_dir_path)
+
+        # Load Metadata
+        return DataSet(files=datafile_paths, metadata=metadata)
+
+    def put(self, item: DataSet) -> None:
+        if self._fs.anon:
+            raise DataStoreException("Cannot perform write operations in anonymous mode. "
+                                     "Please provide credentials with write permissions")
+
+        if not item.metadata:
+            raise DataStoreException("metadata is required in order to persist datasets")
+
+        self._metadata_store.put(item.metadata)
+
+        for filename in item.files:
+            key = self._generate_s3_key(item.id) + filename.name
+            LOGGER.debug(f"Uploading datafile: {filename} to {key}")
+            try:
+                self._fs.put(str(filename), key)
+            except PermissionError:
+                raise DataStoreException(
+                    f"You do not have permission to upload to s3://{key}."
+                    " Please check your credentials are correct or contact the system administrator")
+
+
+class S3FSMetadataStore(BaseS3FSDataStore):
+    """Encapsulates access to stored metadata"""
+
+    def __init__(self, fs: S3FileSystem, storage_bucket_name: str = "caf-data", cache_dir: Optional[Path] = None,
+                 metadata_serialiser: Optional[MetadataSerialiser] = None):
+        self._metadata_serialiser = metadata_serialiser if metadata_serialiser else MetadataYamlSerialiser()
+        super().__init__(fs=fs, storage_bucket_name=storage_bucket_name, cache_dir=cache_dir)
+
+    def _to_s3_key(self, dataset_id: str) -> str:
+        return f"{self._storage_bucket_name}/{dataset_id}/{dataset_id}.metadata"
+
+    def _to_cached_file_path(self, dataset_id: str) -> Path:
+        return (self._cache_dir / Path(dataset_id) / Path(f"{dataset_id}.metadata")).absolute()
+
+    def get(self, dataset_id: str) -> Metadata:
+        metadata_key = self._to_s3_key(dataset_id)
+        cached_metadata_file = self._to_cached_file_path(dataset_id)
+
+        try:
+            self._fs.get(metadata_key, str(cached_metadata_file))
+        except PermissionError:
+            msg = f"PermissionError when accessing {metadata_key}."
+            msg += " Object may not exist, or you may have incorrect/misconfigured credentials"
```

Comment:
I didn't know you could do this `+=` things with error messages, I like it.  And also your very clear error messages, I like those too.

Reply:
yea, python's operator overloading allows type-appropriate behaviour to be implemented. Similar behaviour is implemented for lists as well, for example.

You can actually add bespoke behaviour for any of python's operators by implementing "special" methods on your classes, such as `__add__` to implement the `+` operator.

I believe pretty much any operator can be overloaded. https://www.programiz.com/python-programming/operator-overloading is a reasonable intro, and the definitive reference is https://docs.python.org/3/reference/datamodel.html#special-method-names

I think it's most common to implement "to string", equality, and comparison methods though using things like `__str__`, `__eq__`, `__lt__`, `__lte__` for example